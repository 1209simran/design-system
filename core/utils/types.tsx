export type BaseProps = {
  className?: string,
  'data-test'?: string,
};

const baseProps = [
  'className',
  'data-test'
];

export const extractBaseProps = (props: Record<string, any>) => {
  const basePropsObj = baseProps.reduce((acc, curr) => {
    return (
      props[curr] ? { ...acc, [curr]: props[curr] } : { ...acc }
    );
  }, {});

  return basePropsObj;
};
