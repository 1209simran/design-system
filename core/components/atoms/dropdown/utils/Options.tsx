export let dropdownOptions: any[] = [];
export const group1: any[] = [];
const group2: any[] = [];

for (let i = 1; i <= 40; i++) {
  group1.push({
    label: `Option ${i}`,
    value: `Option ${i}`,
  });
}

for (let i = 41; i <= 50; i++) {
  group2.push({
    label: `Option ${i}`,
    value: `Option ${i}`,
  });
}

dropdownOptions.push({
  group: true,
  label: 'Group 1',
  items: group1,
});

dropdownOptions.push({
  group: true,
  label: 'Group 2',
  items: group2,
});

export const selectedStoryOptions: any[] = [];

for (let i = 1; i <= 10; i++) {
  selectedStoryOptions.push({
    label: `Option ${i}`,
    value: `Option ${i}`,
    selected: i === 2,
  });
}

export const multiSelectedStoryOptions: any[] = [];

for (let i = 1; i <= 10; i++) {
  multiSelectedStoryOptions.push({
    label: `Option ${i}`,
    value: `Option ${i}`,
    selected: i === 2 || i === 7,
  });
}

export const storyOptions = group1.slice(0, 10);

export const storyWrapOptions = [
  {
    label: 'Design System Dropdown',
    value: 'Design System Dropdown'
  },
  {
    label: 'UI Kit Dropdown',
    value: 'UI Kit Dropdown',
  },
  {
    label: 'Innovaccer Analytics',
    value: 'Innovaccer Analytics'
  }
];

export const subInfoItems = [
  {
    label: 'Option 1',
    value: 'Option 1',
    subInfo: 'subInfo'
  },
  {
    label: 'Option 2',
    value: 'Option 2',
    subInfo: 'subInfo'
  },
  {
    label: 'Option 3',
    value: 'Option 3',
    subInfo: 'subInfo',
  },
];

export const iconItems = [
  {
    label: 'Option 1',
    value: 'Option 1',
    icon: 'events'
  },
  {
    label: 'Option 2',
    value: 'Option 2',
    icon: 'events',
  },
  {
    label: 'Option 3',
    value: 'Option 3',
    icon: 'events',
  },
];
