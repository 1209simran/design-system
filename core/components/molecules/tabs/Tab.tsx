import * as React from 'react';

export interface ITabProps {
  label: React.ReactElement;
}

const Tab: React.FunctionComponent<ITabProps> = props => {
  const {
    children,
  } = props;

  return (
    <React.Fragment>
      {children}
    </React.Fragment>
  );
};

export default Tab;
