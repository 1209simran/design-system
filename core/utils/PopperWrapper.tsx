import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Manager, Reference, Popper } from 'react-popper';

type PositionType =
  | 'auto-start'
  | 'auto'
  | 'auto-end'
  | 'top-start'
  | 'top'
  | 'top-end'
  | 'right-start'
  | 'right'
  | 'right-end'
  | 'bottom-end'
  | 'bottom'
  | 'bottom-start'
  | 'left-end'
  | 'left'
  | 'left-start'
  | undefined;

type actionType = 'click' | 'hover';

enum Offsets {
  Small = '2px',
  Medium = '4px',
  Large = '8px'
}

interface IProps {
  trigger: React.ReactElement<any>;
  placement: PositionType;
  children: React.ReactElement<any>;
  style: React.CSSProperties;
  appendToBody: boolean;
  on?: actionType;
  offset: keyof typeof Offsets;
  closeOnBackdropClick?: boolean;
  hoverable?: boolean;
}

interface IState {
  open: boolean;
  mouseLeaveDelay: number;
  mouseEnterDelay: number;
}

class PopperWrapper extends React.Component<IProps, IState> {
  private triggerRef: React.RefObject<HTMLElement>;
  private popupRef: React.RefObject<HTMLDivElement>;
  private _timer?: number;

  constructor(props: IProps) {
    super(props);

    this.state = {
      open: false,
      mouseLeaveDelay: 50,
      mouseEnterDelay: 0
    };

    this.triggerRef = React.createRef();
    this.popupRef = React.createRef();
  }

  public componentWillUnmount() {
    clearTimeout(this._timer);
    document.removeEventListener('mousedown', this.doesNodeContainClick);
  }

  public handleMouseLeave = (event: React.MouseEvent<HTMLElement>) => {
    const { hoverable = false } = this.props;

    if (hoverable) {
      clearTimeout(this._timer);
      this._timer = window.setTimeout(() => {
        this.setState({ open: false });
        if (this.props.children.props.onMouseLeave) {
          this.props.children.props.onMouseLeave(event);
        }
      }, this.state.mouseLeaveDelay);
    } else {
      this.setState({ open: false });
      if (this.props.children.props.onMouseLeave) {
        this.props.children.props.onMouseLeave(event);
      }
    }
  }

  public handleMouseEnter = (event: React.MouseEvent<HTMLElement>) => {
    const { hoverable = false } = this.props;

    if (hoverable) {
      clearTimeout(this._timer);
      this._timer = window.setTimeout(() => {
        this.setState({ open: true });
        if (this.props.children.props.onMouseEnter) {
          this.props.children.props.onMouseEnter(event);
        }
      }, this.state.mouseEnterDelay);
    } else {
      this.setState({ open: true });
      if (this.props.children.props.onMouseEnter) {
        this.props.children.props.onMouseEnter(event);
      }
    }
  }

  public togglePopper = () => {
    this.setState({ open: !this.state.open });
  }

  public doesNodeContainClick = (event: Event) => {
    if (
      !(
        this.findDOMNode(this.popupRef).contains(event.target as HTMLElement) ||
        this.findDOMNode(this.triggerRef).contains(event.target as HTMLElement)
      )
    ) {
      this.togglePopper();
    }
  }

  public componentDidMount() {
    const { on = 'click', closeOnBackdropClick = true } = this.props;
    const { open } = this.state;

    if (on === 'click' && open && closeOnBackdropClick) {
      document.addEventListener('mousedown', this.doesNodeContainClick);
    }
  }

  public componentDidUpdate() {
    const { on = 'click', closeOnBackdropClick = true } = this.props;
    const { open } = this.state;

    if (on === 'click' && open && closeOnBackdropClick) {
      document.addEventListener('mousedown', this.doesNodeContainClick);
    } else if (on === 'click' && !open && closeOnBackdropClick) {
      document.removeEventListener('mousedown', this.doesNodeContainClick);
    }
  }

  public getTriggerElement(trigger: React.ReactElement<any>, ref: React.Ref<any>, on: actionType) {
    const options =
      on === 'hover'
        ? {
          ref,
          onMouseEnter: this.handleMouseEnter,
          onMouseLeave: this.handleMouseLeave
        }
        : {
          ref,
          onClick: this.togglePopper
        };

    const element = React.cloneElement(
      (
        <span className="d-inline-block">
          {trigger}
        </span>
      ),
      options
      );

    return element;
  }

  public getChildrenElement(
    children: React.ReactElement<any>,
    ref: React.Ref<any>,
    placement: string,
    style: React.CSSProperties
  ) {
    const options = this.props.on === 'hover' ? {
      ref,
      style,
      onMouseEnter: this.handleMouseEnter,
      onMouseLeave: this.handleMouseLeave,
      'data-placement': placement
    } : {
      ref,
      style,
      'data-placement': placement
    };

    const element = React.cloneElement(children, options);
    return element;
  }

  public render() {
    const { trigger, children, placement, appendToBody, on = 'click', offset } = this.props;
    const { open } = this.state;

    return (
      <Manager>
        <Reference innerRef={this.triggerRef}>{({ ref }) => this.getTriggerElement(trigger, ref, on)}</Reference>
        {open &&
          appendToBody &&
          ReactDOM.createPortal(
            (
              /* tslint:disable:no-shadowed-variable */
              <Popper placement={placement} innerRef={this.popupRef}>
                {({ ref, style, placement }) => {
                  const newStyle = offset ? this.getUpdatedStyle(style, placement, offset) : style;
                  return this.getChildrenElement(children, ref, placement, newStyle);
                }}
              </Popper>
            ),
            document.body
          )}
        {open && !appendToBody && (
          <Popper placement={placement} innerRef={this.popupRef}>
            {({ ref, style, placement }) => {
              const newStyle = offset ? this.getUpdatedStyle(style, placement, offset) : style;
              return this.getChildrenElement(children, ref, placement, newStyle);
            }}
          </Popper>
        )}
      </Manager>
    );
  }

  private getUpdatedStyle = (
    oldStyle: React.CSSProperties,
    placement: PositionType,
    offset: keyof typeof Offsets = 'Medium'
  ) => {
    const { style = {} } = this.props;
    const newStyle = { ...style, ...oldStyle };
    const position = placement ? placement.split('-')[0] : placement;
    switch (position) {
      case 'top':
        newStyle.marginBottom = Offsets[offset];
        break;

      case 'bottom':
        newStyle.marginTop = Offsets[offset];
        break;

      case 'left':
        newStyle.marginRight = Offsets[offset];
        break;

      case 'right':
        newStyle.marginLeft = Offsets[offset];
        break;
    }
    return newStyle;
  }

  private findDOMNode = (ref: React.RefObject<HTMLElement>) => {
    return ReactDOM.findDOMNode(ref.current!) as HTMLElement;
  }
}

export default PopperWrapper;
