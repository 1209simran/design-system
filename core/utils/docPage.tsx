// @ts-nocheck

import * as React from 'react';
import { Title, Props, Description, Primary, Story, PropsProps } from '@storybook/addon-docs/blocks';

interface DocPageProps {
  title?: string;
  description?: React.ReactNode;
  props?: PropsProps;
}

export const docPage = (props: DocPageProps) => {
  const {
    title,
    description,
    props,
    stories
  } = props;

  const separatorIndex = title?.indexOf('|');

  return (
    <>
      <Title>{title && separatorIndex !== -1 ? title.slice(separatorIndex + 1) : title}</Title>
      <Description>
        {description}
      </Description>
      {/* <Primary /> */}
      {/* <Story id="switch-state--alert"  /> */}
      <Props {...props} />
    </>
  );
};

export default docPage;
