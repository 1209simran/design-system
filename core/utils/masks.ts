import { Mask } from '@/utils/types';
import { DatePickerProps } from "@/index.type";

type DateFormat = DatePickerProps['inputFormat'];
export const date: Record<DateFormat, Mask> = {
  'dd/mm/yyyy': [/[0123]/, /\d/, '/', /\[01]/, /\d/, '/', /\d/, /\d/, /\d/, /\d/],
  'mm/dd/yyyy': [/[01]/, /\d/, '/', /[0123]/, /\d/, '/', /\d/, /\d/, /\d/, /\d/],
  'yyyy/mm/dd': [/\d/, /\d/, /\d/, /\d/, '/', /[01]/, /\d/, '/', /[0123]/, /\d/],
  'dd-mm-yyyy': [/[0123]/, /\d/, '-', /[01]/, /\d/, '-', /\d/, /\d/, /\d/, /\d/],
  'mm-dd-yyyy': [/[01]/, /\d/, '-', /[0123]/, /\d/, '-', /\d/, /\d/, /\d/, /\d/],
  'yyyy-mm-dd': [/\d/, /\d/, /\d/, /\d/, '-', /[01]/, /\d/, '-', /[0123]/, /\d/],
};
