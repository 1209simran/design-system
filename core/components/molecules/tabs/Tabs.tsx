import * as React from 'react';
import classNames from 'classnames';
import { ITabProps } from './Tab';

export interface ITabsProps {
  activeTab?: number;
  children: React.FunctionComponentElement<ITabProps>[];
  onTabChange?: (tabIndex: number) => void;
}

const Tabs: React.FunctionComponent<ITabsProps> = props => {
  const {
    children = [],
    onTabChange
  } = props;

  const [activeTab, setActiveTab] = React.useState(props.activeTab && props.activeTab < children.length
    ? props.activeTab
    : 0);

  React.useEffect(() => {
    setActiveTab(
      props.activeTab && props.activeTab < children.length
      ? props.activeTab
      : 0
    );
  }, [props.activeTab]);

  const wrapperClass = classNames({
    ['Tabs']: true,
  });

  const tabClickHandler = (tabIndex: number) => {
    setActiveTab(tabIndex);
    if (onTabChange) onTabChange(tabIndex);
  };

  const TabsHeader = (
    children.map((child, index) => {
      const tabHeaderClass = classNames({
        ['Tab']: true,
        ['Tab--active']: activeTab === index
      });

      const { label } = child.props;
      return (
        <div key={index} className={tabHeaderClass} onClick={() => tabClickHandler(index)}>
          {label}
        </div>
      );
    })
  );

  return (
    <div className={wrapperClass}>
      <div className="Tabs-header">
        {TabsHeader}
      </div>
      <div className="Tabs-content">
        {children[activeTab]}
      </div>
    </div>
  );
};

export default Tabs;
