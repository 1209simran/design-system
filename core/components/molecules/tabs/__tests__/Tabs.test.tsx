import * as React from 'react';
import { shallow } from 'enzyme';
import { TestHelper, filterUndefined, valueHelper, testMessageHelper } from '@/utils/TestHelper';
import Tab from '../Tab';
import Tabs, { ITabsProps as IProps } from '../Tabs';

const TabValue = [0, 1, 2];
const FunctionValue = [jest.fn()];
const ReactNodeValue = (
  <>
    <Tab label={<></>} />
    <Tab label={<></>} />
  </>
);

const Mapper: Record<string, any> = {
  activeTab: valueHelper(TabValue, { iterate: true }),
  onTabChange: valueHelper(FunctionValue, { iterate: true }),
  children: valueHelper(ReactNodeValue, { required: true }),
};

describe('Tabs component', () => {
  const testFunc = (props: Record<string, any>): void => {
    const attr = filterUndefined(props) as IProps;

    it(testMessageHelper(attr), () => {
      const tree = shallow(
        <Tabs
          {...attr}
        >
          <Tab label={<></>} />
          <Tab label={<></>} />
        </Tabs>
      );
      expect(tree).toMatchSnapshot();
    });
  };

  TestHelper(Mapper, testFunc);
});
