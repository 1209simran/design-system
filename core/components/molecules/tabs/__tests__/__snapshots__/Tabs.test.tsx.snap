// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs component 
  children: "[ReactNode]"
 1`] = `
<div
  className="Tabs"
>
  <div
    className="Tabs-header"
  >
    <div
      className="Tab Tab--active"
      key="0"
      onClick={[Function]}
    />
    <div
      className="Tab"
      key="1"
      onClick={[Function]}
    />
  </div>
  <div
    className="Tabs-content"
  >
    <Tab
      label={<React.Fragment />}
    />
  </div>
</div>
`;

exports[`Tabs component 
  children: "[ReactNode]", activeTab: 0
 1`] = `
<div
  className="Tabs"
>
  <div
    className="Tabs-header"
  >
    <div
      className="Tab Tab--active"
      key="0"
      onClick={[Function]}
    />
    <div
      className="Tab"
      key="1"
      onClick={[Function]}
    />
  </div>
  <div
    className="Tabs-content"
  >
    <Tab
      label={<React.Fragment />}
    />
  </div>
</div>
`;

exports[`Tabs component 
  children: "[ReactNode]", activeTab: 0, onTabChange: "[Function]"
 1`] = `
<div
  className="Tabs"
>
  <div
    className="Tabs-header"
  >
    <div
      className="Tab Tab--active"
      key="0"
      onClick={[Function]}
    />
    <div
      className="Tab"
      key="1"
      onClick={[Function]}
    />
  </div>
  <div
    className="Tabs-content"
  >
    <Tab
      label={<React.Fragment />}
    />
  </div>
</div>
`;

exports[`Tabs component 
  children: "[ReactNode]", activeTab: 1
 1`] = `
<div
  className="Tabs"
>
  <div
    className="Tabs-header"
  >
    <div
      className="Tab"
      key="0"
      onClick={[Function]}
    />
    <div
      className="Tab Tab--active"
      key="1"
      onClick={[Function]}
    />
  </div>
  <div
    className="Tabs-content"
  >
    <Tab
      label={<React.Fragment />}
    />
  </div>
</div>
`;

exports[`Tabs component 
  children: "[ReactNode]", activeTab: 1, onTabChange: "[Function]"
 1`] = `
<div
  className="Tabs"
>
  <div
    className="Tabs-header"
  >
    <div
      className="Tab"
      key="0"
      onClick={[Function]}
    />
    <div
      className="Tab Tab--active"
      key="1"
      onClick={[Function]}
    />
  </div>
  <div
    className="Tabs-content"
  >
    <Tab
      label={<React.Fragment />}
    />
  </div>
</div>
`;

exports[`Tabs component 
  children: "[ReactNode]", activeTab: 2
 1`] = `
<div
  className="Tabs"
>
  <div
    className="Tabs-header"
  >
    <div
      className="Tab Tab--active"
      key="0"
      onClick={[Function]}
    />
    <div
      className="Tab"
      key="1"
      onClick={[Function]}
    />
  </div>
  <div
    className="Tabs-content"
  >
    <Tab
      label={<React.Fragment />}
    />
  </div>
</div>
`;

exports[`Tabs component 
  children: "[ReactNode]", activeTab: 2, onTabChange: "[Function]"
 1`] = `
<div
  className="Tabs"
>
  <div
    className="Tabs-header"
  >
    <div
      className="Tab Tab--active"
      key="0"
      onClick={[Function]}
    />
    <div
      className="Tab"
      key="1"
      onClick={[Function]}
    />
  </div>
  <div
    className="Tabs-content"
  >
    <Tab
      label={<React.Fragment />}
    />
  </div>
</div>
`;

exports[`Tabs component 
  children: "[ReactNode]", onTabChange: "[Function]"
 1`] = `
<div
  className="Tabs"
>
  <div
    className="Tabs-header"
  >
    <div
      className="Tab Tab--active"
      key="0"
      onClick={[Function]}
    />
    <div
      className="Tab"
      key="1"
      onClick={[Function]}
    />
  </div>
  <div
    className="Tabs-content"
  >
    <Tab
      label={<React.Fragment />}
    />
  </div>
</div>
`;
